### PR TITLE
EVG-6097: add configuration options for communication with hosts

### PR DESCRIPTION
**Pull request by Kimchelly**
JIRA: https://jira.mongodb.org/browse/EVG-6097

* Add option to configure how the server communicates with hosts (legacy, ssh, RPC).

### CODE REVIEW

File: model/distro/distro.go [hunk 1]
```diff
@@ -252,6 +265,29 @@ func ValidateBootstrapMethod(method string) error {
...
+}
+
+// ValidateBootstrapAndCommunicationMethods checks that the bootstrap and
+// communication mechanisms are recognized, and that the combination of the two
+// is allowed.
+func ValidateBootstrapAndCommunicationMethods(bootstrap string, communication string) error {
+	catcher := grip.NewBasicCatcher()
+	catcher.Add(ValidateBootstrapMethod(bootstrap))
+	catcher.Add(ValidateCommunicationMethod(communication))
+	if bootstrap == BootstrapMethodLegacySSH && communication != CommunicationMethodLegacySSH ||
```
Comment (tychoish) [inline model/distro/distro.go]: I'd use a switch statement rather than the OR clause

File: model/distro/distro.go [hunk 2]
```diff
@@ -100,6 +106,13 @@ var ValidBootstrapMethods = []string{
 	BootstrapMethodUserData,
 }
 
+// ValidCommunicationMethods includes all recognized host communication methods.
+var ValidCommunicationMethods = []string{
```
Comment (tychoish) [inline model/distro/distro.go]: I think that this can be private because we're only really interacting with it via the public method.

File: validator/distro_validator_test.go [hunk 3]
```diff
@@ -360,16 +364,25 @@ func TestEnsureValidContainerPool(t *testing.T) {
...
-		distro.BootstrapMethodLegacySSH,
-		distro.BootstrapMethodSSH,
-		distro.BootstrapMethodPreconfiguredImage,
-		distro.BootstrapMethodUserData,
-	} {
-		assert.Nil(t, ensureValidBootstrapMethod(ctx, &distro.Distro{BootstrapMethod: bootstrapMethod}, &evergreen.Settings{}))
+	for _, bootstrapMethod := range distro.ValidBootstrapMethods {
+		for _, communicationMethod := range distro.ValidCommunicationMethods {
+			d := &distro.Distro{BootstrapMethod: bootstrapMethod, CommunicationMethod: communicationMethod}
+			if bootstrapMethod == distro.BootstrapMethodLegacySSH && communicationMethod == distro.CommunicationMethodLegacySSH {
```
Comment (tychoish) [inline validator/distro_validator_test.go]: switch statement? this logic. seems a bit hard to follow.

File: model/distro/distro.go (hunk 2 above)
Comment (Kimchelly) [reply model/distro/distro.go]: Changed to unexported.

File: model/distro/distro.go (hunk 1 above)
Comment (Kimchelly) [reply model/distro/distro.go]: Done.

File: validator/distro_validator_test.go (hunk 3 above)
Comment (Kimchelly) [reply validator/distro_validator_test.go]: Done.

File: model/distro/distro.go [hunk 4]
```diff
@@ -281,9 +281,15 @@ func ValidateBootstrapAndCommunicationMethods(bootstrap string, communication st
 	catcher := grip.NewBasicCatcher()
 	catcher.Add(ValidateBootstrapMethod(bootstrap))
 	catcher.Add(ValidateCommunicationMethod(communication))
-	if bootstrap == BootstrapMethodLegacySSH && communication != CommunicationMethodLegacySSH ||
-		communication == CommunicationMethodLegacySSH && bootstrap != BootstrapMethodLegacySSH {
-		catcher.Add(fmt.Errorf("'%s' and '%s' is not a valid bootstrap and communication combination - legacy SSH is incompatible with non-legacy functionality", bootstrap, communication))
+	switch bootstrap {
+	case BootstrapMethodLegacySSH:
+		if communication != CommunicationMethodLegacySSH {
+			catcher.Add(fmt.Errorf("bootstrapping hosts using legacy SSH is incompatible with non-legacy host communication"))
```
Comment (tychoish) [inline model/distro/distro.go]: you only need to use Errorf if you're doing something sprintf-ish, use errors.New() otherwise. 

(also in general prefer errors.Errorf from pkg/errors over fmt.Errorf or) 

having said that, the catcher interface now, has a .New() method that takes a string.
